### ISSUE TITLE
api for melpa

### ISSUE THREAD

**Issue by ChillarAnand** [milkypostman/melpa#2850]
for 3rd parties to interact with melpa, there is only archives.json

if melpa provides a simple json api to search & get information about packages, it would be lot better

**Comment by milkypostman** [comment]
Except we don't actually run anything more than a static server. And I'm
not sure we want to do more. Cause maintaining some server is a lot of
work. What is the problem with the JSON?

**Comment by ChillarAnand** [comment]
i am building a goodie for duckduckgo. when someone seaches for `emacs package powerline` or something like that, melpa info will be displayed at the top.
![mel](https://cloud.githubusercontent.com/assets/4463796/8143169/67ecd102-11bc-11e5-9a60-80a81975b5c6.png)


duckduckgo team is not ok with archives.json, as it is too large(>500kb).

if you could add one more view which takes package name as input and give info as json object, that would be helpful?

**Comment by purcell** [comment]
It would be possible, but this is likely to be a very low priority for us, because we don't currently have any sort of long-running server side component.

**Comment by milkypostman** [comment]
we could generate a json file per package. would that be enough?

**Comment by ChillarAnand** [comment]
@milkypostman awesome. that will do the job.

**Comment by benjamreynolds** [comment]
Is this still on the roadmap? I am working on something similar for [Octolinker](https://github.com/OctoLinker/browser-extension) That would really benefit from this.

If this needs someone to step in and work on it, Im happy dig in and see what it would require.

**Comment by milkypostman** [comment]
i worked on it a bit but the data required to track this and the CPU to regenerate were quite large. I am still working on a solution.

**Comment by milkypostman** [comment]
i'm open to collaboration or contributions.

**Comment by alphapapa** [comment]
> i worked on it a bit but the data required to track this and the CPU to regenerate were quite large

What if `package-build` were to output a JSON file when it built a package?  Maybe a MELPA-specific function could advise it, so it would only run on the MELPA server.

**Comment by purcell** [comment]
> What if package-build were to output a JSON file when it built a package? Maybe a MELPA-specific function could advise it, so it would only run on the MELPA server.

More generally, I'd like `package-build.el` to have hooks that it would call after building packages, passing them info about the packages. This would allow us to move MELPA-specific stuff out of that library.

**Comment by alphapapa** [comment]
Would passing a plist be good, so it would be more future-proofed?

Sorry if this is getting too specific.  I thought maybe I could help implement it, although no promises, of course.  ;)  Thanks.

**Comment by josephfrazier** [comment]
Inspired by @benjamreynolds, I adapted https://github.com/OctoLinker/live-resolver/pull/41 into an independent service: https://melpa.glitch.me. Here's an example using [httpie](https://github.com/jakubroztocil/httpie):

```bash
$ http https://melpa.glitch.me/parinfer
HTTP/1.1 200 OK
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Mon, 19 Jun 2017 16:27:22 GMT
Transfer-Encoding: chunked
cache-control: no-cache
content-encoding: gzip
vary: accept-encoding
x-powered-by: Express

{
    "deps": {
        "cl-lib": [
            0,
            5
        ],
        "dash": [
            2,
            13,
            0
        ]
    },
    "desc": "Simpler Lisp editing",
    "props": {
        "commit": "972d413b4d8284ba67ab43513a6a709692325f20",
        "keywords": [
            "parinfer"
        ],
        "url": "https://github.com/DogLooksGood/parinfer-mode"
    },
    "type": "tar",
    "ver": [
        20170612,
        1048
    ]
}
```

You can see/take the code here: https://glitch.com/edit/#!/melpa?path=server.js

**Comment by alphapapa** [comment]
@josephfrazier Cool.  It doesn't cache the JSON file though, right?
```js
const response = await got('https://melpa.org/archive.json', {json: true});
```

**Comment by josephfrazier** [comment]
Correct, but you can add caching in by keeping track of the `Last-Modified` header and the previous response outside the request handler:

```js
let lastModified;
let archive;
```

then using those inside the request handler:

```js
  try {
    const response = await got('https://melpa.org/archive.json', {
      json: true,
      headers: lastModified ? {
        'if-modified-since': lastModified,
      } : undefined,
    });

    lastModified = response.headers['last-modified'];
    archive = response.body;
  } catch (err) {
    if (err.statusCode !== 304) {
      throw err;
    }
  }
```

**Comment by tarsius** [comment]
We provide an "entry" file for each package. It contains the same information as the respective entry in "archive-contents". This will have to do.

The URL of that file is the same as for the respective tarball, except for the suffix. Replace `.tar` with `.entry`.
